### PR TITLE
MODINREACH-98 - Return Uncirculated Message for Item Hold

### PR DESCRIPTION
**Pull request by s1mpleGit**
[MODINREACH-98](https://issues.folio.org/browse/MODINREACH-98) - Return Uncirculated Message for Item Hold

## Purpose
Endpoint receives message from central server to owning site indicating that an item that was checked out to and received by the borrowing site (shipped) is being returned by the borrowing site without being checked out by the requesting patron.

## Approach
Local endpoint implemented
Tests added

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

### CODE REVIEW

File: src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    "itemAgencyCode": {
+      "description": "Item agency code",
+      "type": "string",
+      "pattern": "[a-z,0-9]{5}"
+    },
+    "itemId": {
+      "description": "Item id",
+      "type": "string",
+      "pattern": "[a-z,0-9]{1,32}"
+    },
```
Comment (oliinyko) [inline src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json]: these fields are already defined in baseCircRequestDTO.json.
You can inherit them by adding:
`  "allOf": [
    {
      "$ref": "baseCircRequestDTO.json"
    }
  ]`

Comment (s1mpleGit) [reply src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json]: fixed

**Review by oliinyko — APPROVED** [review]

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java [hunk 2]
```diff
@@ -181,6 +184,20 @@ public InnReachResponseDTO receiveUnshipped(String trackingId, String centralCod
     return success();
   }
 
+  @Override
+  public InnReachResponseDTO returnUncirculated(String trackingId, String centralCode, ReturnUncirculatedDTO returnUncirculated) {
+    var transaction = getTransaction(trackingId, centralCode);
+    InnReachTransaction.TransactionState state = transaction.getState();
```
Comment (maryna-steshenko) [inline src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]:
```suggestion
    var state = transaction.getState();
```

**Review by maryna-steshenko — APPROVED** [review]

File: src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java (hunk 2 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/CirculationServiceImpl.java]: fixed

File: src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  "properties": {
+    "title": {
+      "description": "Item title",
+      "type": "string",
+      "maxLength": 256
+    },
+    "author": {
+      "description": "Author",
+      "type": "string",
+      "maxLength": 256
```
Comment (dmtkachenko) [inline src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json]: there is no restriction on length:
image

Comment (s1mpleGit) [reply src/main/resources/swagger.api/schemas/d2ir/returnUncirculatedDTO.json]: fixed

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 4]
```diff
@@ -291,4 +294,79 @@ void precessReportUnshippedItemReceived_whenTransactionItemShipped() {
     assertNotEquals(RECEIVE_UNANNOUNCED, transactionUpdated.getState());
   }
 
+  @Test
+  @Sql(scripts = {
+    "classpath:db/central-server/pre-populate-central-server.sql",
+    "classpath:db/inn-reach-transaction/pre-populate-inn-reach-transaction.sql"
+  })
+  void checkTransactionIsInStateItemReceived() {
```
Comment (dmtkachenko) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: checkTransactionIsInStateItemReceived() and checkTransactionIsInStateReceiveUnannounced() methods perform the same steps and validation apart from setting the initial state. They can be combined into a single test with different inputs: 
- intial state = ITEM_RECEIVED
- intial state = RECEIVE_UNANNOUNCED

You can use @ParameterizedTest to achieve the above. A couple of samples:
- https://github.com/folio-org/mod-inn-reach/blob/4497d0f40df3183b7a108e362f1787274b3933de/src/test/java/org/folio/innreach/config/props/FolioEnvironmentTest.java#L42
- https://github.com/folio-org/mod-inn-reach/blob/4497d0f40df3183b7a108e362f1787274b3933de/src/test/java/org/folio/innreach/controller/d2ir/TransferRequestCirculationApiTest.java#L72

Comment (s1mpleGit) [reply src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: fixed

**Review by dmtkachenko — APPROVED** [review]

**Pull request merged**